src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,112 @@
-
+        ensureCapacity(size + 1);
+        elements[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Invalid index: " + index);
+        }
+        ensureCapacity(size + 1);
+        System.arraycopy(elements, index, elements, index + 1, size - index);
+        elements[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        if (list == null || list.size() == 0) {
+            return;
+        }
+        ensureCapacity(size + list.size());
+        for (int i = 0; i < list.size(); i++) {
+            elements[size++] = list.get(i);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Invalid index: " + index);
+        }
+        return (T) elements[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Invalid index: " + index);
+        }
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Invalid index: " + index);
+        }
+        T removedElement = (T) elements[index];
+        System.arraycopy(elements, index + 1, elements, index, size - index - 1);
+        elements[--size] = null; // Clear the reference to the removed element.
+        return removedElement;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if (elementsEqual(element, elements[i])) {
+                return remove(i);
+            }
+        }
+        throw new NoSuchElementException("Element not found: " + element);
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void ensureCapacity(int requiredCapacity) {
+        if (requiredCapacity > elements.length) {
+            int newCapacity = (int) (elements.length * GROWTH_FACTOR);
+            if (newCapacity < requiredCapacity) {
+                newCapacity = requiredCapacity;
+            }
+            Object[] newArray = new Object[newCapacity];
+            System.arraycopy(elements, 0, newArray, 0, size);
+            elements = newArray;
+        }
+    }
+
+    private boolean elementsEqual(T element1, Object element2) {
+        if (element1 == null && element2 == null) {
+            return true;
+        }
+        if (element1 == null || element2 == null) {
+            return false;
+        }
+        return element1.equals(element2);
     }
```

Comment:
You could simplify the `elementsEqual()` method by using `Objects.equals(element1, element2)`. This method checks for null and calls `equals()` if both objects are not null. This will make your code cleaner and easier to read.